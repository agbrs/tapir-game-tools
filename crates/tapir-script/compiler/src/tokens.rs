use std::num::ParseIntError;

use logos::Logos;
use num_traits::ParseFloatError;
use serde::Serialize;

#[derive(Clone, Debug, Copy, Hash, PartialEq, Eq, Serialize)]
pub struct FileId(usize);

impl FileId {
    pub fn new(id: usize) -> Self {
        Self(id)
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize)]
pub struct Span {
    pub(crate) file_id: FileId,
    pub(crate) start: usize,
    pub(crate) end: usize,
}

impl Span {
    pub fn new(file_id: FileId, start: usize, end: usize) -> Self {
        assert!(start <= end, "{} was not <= {}", start, end);
        Self {
            file_id,
            start,
            end,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct LexicalError {
    pub kind: LexicalErrorKind,
    pub span: Span,
}

#[derive(Default, Debug, Clone, PartialEq, Eq, Serialize)]
#[allow(clippy::enum_variant_names)]
pub enum LexicalErrorKind {
    InvalidNumber(#[serde(skip)] ParseIntError),
    InvalidFix,
    #[default]
    InvalidToken,
}

impl From<ParseIntError> for LexicalErrorKind {
    fn from(value: ParseIntError) -> Self {
        Self::InvalidNumber(value)
    }
}

impl From<ParseFloatError> for LexicalErrorKind {
    fn from(_: ParseFloatError) -> Self {
        Self::InvalidFix
    }
}

impl LexicalErrorKind {
    pub fn with_span(self, file_id: FileId, start: usize, end: usize) -> LexicalError {
        LexicalError {
            kind: self,
            span: Span::new(file_id, start, end),
        }
    }
}

#[derive(Logos, Clone, Debug, PartialEq, Serialize)]
#[logos(skip r"[ \t\n\f\r]+", skip r"#.*\n?", error = LexicalErrorKind)]
pub enum Token<'input> {
    #[token("wait")]
    KeywordWait,
    #[token("var")]
    KeywordVar,

    #[regex("[_a-zA-Z][_0-9a-zA-Z]*", |lex| lex.slice())]
    Identifier(&'input str),
    #[regex("-?[0-9]+", |lex| lex.slice())]
    Integer(&'input str),
    #[regex("-?[0-9]+\\.[0-9]*", |lex| lex.slice())]
    Fix(&'input str),

    #[token("true")]
    True,
    #[token("false")]
    False,

    #[token("if")]
    If,
    #[token("else")]
    Else,

    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("=")]
    Assign,
    #[token(";")]
    Semicolon,

    #[token("+")]
    OperatorAdd,
    #[token("-")]
    OperatorSub,
    #[token("*")]
    OperatorMul,
    #[token("/")]
    OperatorDiv,
    #[token("%")]
    OperatorMod,
    #[token("//")]
    OperatorRealDiv,
    #[token("%%")]
    OperatorRealMod,
    #[token("==")]
    OperatorEqEq,
}
