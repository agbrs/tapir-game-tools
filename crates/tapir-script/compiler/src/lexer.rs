use logos::{Logos, SpannedIter};

use crate::tokens::{FileId, LexicalError, Token};

pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;

pub struct Lexer<'input> {
    token_stream: SpannedIter<'input, Token<'input>>,
    file_id: FileId,
}

impl<'input> Lexer<'input> {
    pub fn new(input: &'input str, file_id: FileId) -> Self {
        Self {
            token_stream: Token::lexer(input).spanned(),
            file_id,
        }
    }
}

impl<'input> Iterator for Lexer<'input> {
    type Item = Spanned<Token<'input>, usize, LexicalError>;

    fn next(&mut self) -> Option<Self::Item> {
        let (token, span) = self.token_stream.next()?;

        Some(match token {
            Ok(token) => Ok((span.start, token, span.end)),
            Err(err) => Err(err.with_span(self.file_id, span.start, span.end)),
        })
    }
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_ron_snapshot, glob};

    use super::*;

    #[test]
    fn snapshot_tests() {
        glob!("snapshot_tests", "lexer/*.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let output = lexer
                .map(|token| token.map(|(_, token, _)| token))
                .collect::<Vec<_>>();

            assert_ron_snapshot!(output);
        });
    }
}
