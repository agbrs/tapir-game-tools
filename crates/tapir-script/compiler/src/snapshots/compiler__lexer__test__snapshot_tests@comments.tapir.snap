---
source: crates/tapir-script/compiler/src/lexer.rs
expression: output
input_file: crates/tapir-script/compiler/src/snapshot_tests/lexer/comments.tapir
---
[
  Ok(KeywordWait),
  Ok(Semicolon),
  Ok(KeywordWait),
  Ok(Semicolon),
]
